PubSub Manager, aka Qless-native Events

stephenreay opened the pull request:
* Type: new feature
* Link to issue: #100 

**In raising this pull request, I confirm the following (please check boxes):**

- [x] I have read and understood the [Contributing Guidelines](https://github.com/pdffiller/qless-php/blob/master/CONTRIBUTING.md)
- [x] I have checked that another pull request for this purpose does not exist.
- [x] I wrote some tests for this PR.


This roughly follows the Python (and to a lesser extent Ruby) API, where `Client` exposes an `events` property which has `on(name, callback)` and `listen()` methods.

I noticed the existing Watchdog PubSub consumer does a clone/disconnect/reconnect, but that method does not seem to reliably give a new connection (to allow the callback to use `Client` for other calls).

The current implementation also doesn't offer the priority functionality of the existing callback based "Events" system - I thought it more important to keep parity with the upstream feature.

I had to make a guess about naming too, it's not clear to me whether this class should be in the root `Qless` namespace, or even if you have some better name than `PubSub Manager` - EventsManager is taken and I assume you wouldn't just drop the old implementation.

I've done some very minimal tests of this and it works as expected across processes. It'd be great to get some feedback in terms of naming/structure/etc and I can refine it from there.

stephenreay committed:
Initial working example for a PubSub Manager, aka Qless-native Events

This roughly follows the Python (and to a lesser extent Ruby) API, where `Client` exposes an `events` property which has `on(name, callback)` and `listen()` methods.

I noticed the existing Watchdog PubSub consumer does a clone/disconnect/reconnect, but that method does not seem to reliably give a new connection (to allow the callback to use `Client` for other calls).

The current implementation also doesn't offer the priority functionality of the existing callback based "Events" system - I thought it more important to keep parity with the upstream feature.

I had to make a guess about naming too, it's not clear to me whether this class should be in the root `Qless` namespace, or even if you have some better name than `PubSub Manager` - EventsManager is taken and I assume you wouldn't just drop the old implementation.

I've done some very minimal tests of this and it works as expected across processes. It'd be great to get some feedback in terms of naming/structure/etc and I can refine it from there.

stephenreay committed:
Updated README

stephenreay committed:
Fixed formatting for PR

osadchyi-s commented:
Hi. Cool PR. What status of this pr? I see _WIP_ . 
If you will add some tests - it will be great.

stephenreay committed:
Updated PubSub Event Manager

Don't take unused Client in PubSub\Manager constructor
Add a `stopListening` method to PubSub\Manager
Expand README with list of notification event names

stephenreay committed:
Add tests for PubSub Notifications system

Add a simple test worker that doesn't fork
Add helper script to capture and return event notifications in the background
Add DummyPubSubJob to trigger various event notifications
Add Test case for PubSub\Manager

stephenreay committed:
Use a fixed path to PHP binary

stephenreay commented:
Hi @osadchyi-s 

Sorry for the delay getting back to this.

I've updated the API slightly, filled out the README a little more, and added tests.

Cheers

Stephen

stephenreay committed:
Friends don't let friends shift too many elements off an array.

stephenreay committed:
Try using relative `php` for the tests.

Also outputs test logs on failure.

stephenreay committed:
Fixes to get the tests green.

stephenreay commented:
Hey @osadchyi-s any chance you could take another look at this?
